File: rust/src/bin/hello.rs
use async_openai::types::{
    ChatCompletionRequestMessage, ChatCompletionRequestUserMessageArgs,
    CreateChatCompletionRequestArgs,
};
use rein_ch01_helloworld::{build_client, load_config, AppError};
use std::time::Duration;

#[tokio::main]
async fn main() {
    dotenvy::dotenv().ok();
    let result = async {
        let config = load_config()?;
        let client = build_client(&config);
        let message = ChatCompletionRequestUserMessageArgs::default()
            .content("hello")
            .build()
            .map_err(|_| AppError::Sdk)?;
        let request = CreateChatCompletionRequestArgs::default()
            .model(&config.model)
            .messages([ChatCompletionRequestMessage::User(message)])
            .build()
            .map_err(|_| AppError::Sdk)?;
        let response = tokio::time::timeout(Duration::from_secs(30), client.chat().create(request))
            .await
            .map_err(|_| AppError::Timeout)?
            .map_err(|_| AppError::Network)?;
        response
            .choices
            .first()
            .and_then(|choice| choice.message.content.as_deref())
            .filter(|text| !text.trim().is_empty())
            .ok_or(AppError::ResponseFormat)
            .map(str::to_owned)
    }
    .await;
    match result {
        Ok(text) => println!("{text}"),
        Err(_error) => {
            eprintln!("调用失败，请检查配置、网络和服务端响应；需要分类诊断时运行安全入口");
            std::process::exit(1);
        }
    }
}

File: ts/src/hello.ts
import OpenAI, { type ClientOptions } from 'openai'
import { pathToFileURL } from 'node:url'

import type { Config } from './config'

export const HELLO_TIMEOUT_MS = 30_000

export interface HelloOptions {
  readonly prompt?: string
  readonly fetch?: ClientOptions['fetch']
  readonly timeoutMs?: number
}

/** 最小的非流式 Chat Completions 调用，适合第一次把 hello 送到模型。 */
export async function requestHello(
  config: Config,
  options: HelloOptions = {},
): Promise<string> {
  const client = new OpenAI({
    apiKey: config.apiKey,
    baseURL: config.baseUrl,
    maxRetries: 0,
    timeout: HELLO_TIMEOUT_MS,
    ...(options.fetch === undefined ? {} : { fetch: options.fetch }),
  })
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), options.timeoutMs ?? HELLO_TIMEOUT_MS)
  let response: unknown
  try {
    response = await client.chat.completions.create({
      model: config.model,
      messages: [{ role: 'user', content: options.prompt ?? 'hello' }],
      stream: false,
    }, { signal: controller.signal })
  } finally {
    clearTimeout(timer)
  }
  const text = readHelloText(response)
  if (typeof text !== 'string' || text.trim().length === 0) {
    throw new Error('模型返回了空的回答文本。')
  }
  return text
}

export async function main(argv: readonly string[] = process.argv.slice(2)): Promise<number> {
  const { loadConfig } = await import('./config')
  try {
    const text = await requestHello(loadConfig(), { prompt: argv.join(' ').trim() || 'hello' })
    console.log(text)
    return 0
  } catch (error) {
    console.error('调用失败，请运行 hello-safe.ts 查看经过分类的诊断信息。')
    return 1
  }
}

if (process.argv[1] !== undefined && import.meta.url === pathToFileURL(process.argv[1]).href) {
  process.exitCode = await main()
}

function readHelloText(value: unknown): string {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) throw new Error('模型响应不是对象。')
  const choices = (value as { choices?: unknown }).choices
  if (!Array.isArray(choices) || choices.length === 0) throw new Error('模型响应缺少 choices[0]。')
  const first = choices[0]
  if (typeof first !== 'object' || first === null || Array.isArray(first)) throw new Error('模型响应缺少 choices[0].message。')
  const message = (first as { message?: unknown }).message
  if (typeof message !== 'object' || message === null || Array.isArray(message)) throw new Error('模型响应缺少 choices[0].message。')
  const text = (message as { content?: unknown }).content
  if (typeof text !== 'string' || text.trim().length === 0) throw new Error('模型返回了空的回答文本。')
  return text
}
